src/games/strategy/triplea/ai/proAI/util/ProOddsCalculator.java

```diff
@@ -0,0 +1,173 @@
+    final boolean isLandAndCanOnlyBeAttackedByAir = !t.isWater() && Match.allMatch(attackingUnits, Matches.UnitIsAir);
+    if (attackingUnits.size() == 0) {
+      return new ProBattleResult();
+    } else if (hasNoDefenders && isLandAndCanOnlyBeAttackedByAir) {
+      return new ProBattleResult();
+    } else if (hasNoDefenders) {
+      return new ProBattleResult(100, 0.1, true, attackingUnits, new ArrayList<Unit>(), 0);
+    } else if (Properties.getSubRetreatBeforeBattle(data) && Match.allMatch(defendingUnits, Matches.UnitIsSub)
+        && Match.noneMatch(attackingUnits, Matches.UnitIsDestroyer)) {
+      return new ProBattleResult();
+    }
+    return null;
+  }
+
+
+  public ProBattleResult callBattleCalculator(final PlayerID player, final Territory t,
+      final List<Unit> attackingUnits, final List<Unit> defendingUnits, final Set<Unit> bombardingUnits) {
+    return callBattleCalculator(player, t, attackingUnits, defendingUnits, bombardingUnits, false);
+  }
+
+  public ProBattleResult callBattleCalculator(final PlayerID player, final Territory t,
+      final List<Unit> attackingUnits, final List<Unit> defendingUnits, final Set<Unit> bombardingUnits,
+      final boolean retreatWhenOnlyAirLeft) {
+    final GameData data = ProData.getData();
+
+    if (isCanceled || attackingUnits.isEmpty() || defendingUnits.isEmpty()) {
+      return new ProBattleResult();
+    }
+
+    // Use battle calculator (hasLandUnitRemaining is always true for naval territories)
+    AggregateResults results = null;
+    final int minArmySize = Math.min(attackingUnits.size(), defendingUnits.size());
+    final int runCount = Math.max(16, 100 - minArmySize);
+    final PlayerID attacker = attackingUnits.get(0).getOwner();
+    final PlayerID defender = defendingUnits.get(0).getOwner();
+    if (retreatWhenOnlyAirLeft) {
+      calc.setRetreatWhenOnlyAirLeft(true);
+    }
+    results =
+        calc.setCalculateDataAndCalculate(attacker, defender, t, attackingUnits, defendingUnits, new ArrayList<Unit>(
+            bombardingUnits), TerritoryEffectHelper.getEffects(t), runCount);
+    if (retreatWhenOnlyAirLeft) {
+      calc.setRetreatWhenOnlyAirLeft(false);
+    }
+
+    // Find battle result statistics
+    final double winPercentage = results.getAttackerWinPercent() * 100;
+    final List<Unit> averageAttackersRemaining = results.GetAverageAttackingUnitsRemaining();
+    final List<Unit> averageDefendersRemaining = results.GetAverageDefendingUnitsRemaining();
+    final List<Unit> mainCombatAttackers =
+        Match.getMatches(attackingUnits, Matches.UnitCanBeInBattle(true, !t.isWater(), data, 1, false, true, true));
+    final List<Unit> mainCombatDefenders =
+        Match.getMatches(defendingUnits, Matches.UnitCanBeInBattle(false, !t.isWater(), data, 1, false, true, true));
+    double TUVswing = results.getAverageTUVswing(attacker, mainCombatAttackers, defender, mainCombatDefenders, data);
```

Comment:
This code is just moved from ProBattleUtils and not being refactored at this point.